pkg/cosign/verify.go

```diff
@@ -725,15 +725,20 @@ func keyBytes(sig oci.Signature, co *CheckOpts) ([]byte, error) {
 	if err != nil {
 		return nil, err
 	}
-	// We have a public key.
-	if co.SigVerifier != nil {
+	switch {
+	// certificate provided from the OCI image, or via flag (--certificate or --bundle)
```

Comment:
I'm not 100% sure I understand the logic here, but I'm worried that this bit might have a subtle verification bug. Let me know if this makes sense:

If the user has specified `--certificate` at the command line my understanding is that it get parsed and its public key is used in initialize `co.SigVerifier`. This function appears to then return a certificate PEM if found on the OCI signature (which may be different than `--certificate` 😬 ). The function calling this one https://github.com/sigstore/cosign/blob/29360f6a3390d44dd8faef636dd0c3449a213c88/pkg/cosign/verify.go#L1078-L1081 then appears to compare the returned certificate to the one it finds in the rekor bundle.

My worry here is that the certificate compared to the rekor bundle might not be the one the user actually specified at the command line, unless that bit of logic is lurking in a different place?

Reply:
I think we have this issue now also, although it doesn’t surface any error. If you provide a certificate that certified the same public key as a certificate on the OCI imagine, but the rest of the cert differs, we wouldn’t notice. 

Note this isn’t an issue for verify-blob since we construct a static OCI struct from input. For verify on OCI, theres no way to pass the certificate passed in from input  

Sigh…maybe we could add a check that a provided cert matches what’s on the image annotations? Or disallow providing a certificate when the image annotations contain a certificate already?

Reply:
> Sigh…maybe we could add a check that a provided cert matches what’s on the image annotations? Or disallow providing a certificate when the image annotations contain a certificate already?

+1, see my comment below. I think `--certificate` is a little misleading here. The `--certificate` flag for OCI image never "makes" it into the pkg verification code (just the pub key does). In the CLI layer there's some verification of it, but once we get into `pkg/verify`, we rely on `sig.Cert()`.